### ISSUE TITLE
Capturing All Events in Audit Trail

### ISSUE THREAD

**Issue by parkinsona** [IdentityServer/IdentityServer3.MembershipReboot#43]
Hi,

I have a requirement to create an audit trail of all actions that modify the userAccount.
Our initial solution was to add a trigger to the UserAccount table so that we could see a history of what was changing. This was good, however we need to know which MR User is making the change in addition to which DB User is making the change.

I know you can write event handlers for MR, is there a way to write 1 global event handler that will insert into an audit Trail table for all events that happen in MR (Eg Change Password, Login, Verified, Closed, etc)?

**Comment by brockallen** [comment]
One of the sames shows something like this. It might help get you started./

**Comment by parkinsona** [comment]
I'm not sure which sample you're referring to, but here's what I was thinking.

I could override the EmailEventHandler's Process method to include the call to write the Audit Trail entry. This would allow me to do it in 1 place, while retaining the existing functionality of the EmailEventHandler.

The one thing I'm not sure on, is how I could get the currently logged in UserID into the Audit Trail.

**Comment by brockallen** [comment]
Sorry, I thought this was opened in the MR repo, but the sample I was thinking about is here: https://github.com/brockallen/BrockAllen.MembershipReboot/blob/master/samples/CustomizationsSample/CustomizationsSample/Customizations.cs#L130

**Comment by parkinsona** [comment]
Ahhh I see what you've got there. That makes sense.
Is there a list of all events somewhere? 
Does this have all the events used: https://github.com/brockallen/BrockAllen.MembershipReboot/blob/2f1e5d4b39e6f9a567eeca634665a1551615341e/src/BrockAllen.MembershipReboot/AccountService/UserAccountEvents.cs

In that sample I see you're using HttPContext, which means I should be able to retrieve information for the user if they're logged in. For login failures, I obviously can't give them who "failed", but that's fine.

**Comment by brockallen** [comment]
Yep that's all the events. 

-Brock

**Comment by parkinsona** [comment]
I have this kind of working now, but I have 2 questions

1. Is it possible to write a handler for the UserAccountEvent class that will fire for all its children? 
2. In the sample you use:  HttpContext.Current.Request.UserHostAddress. This keeps returning ::1. I tried returning claims from HttpContext, but the claim list was also empty. Am I missing something?

**Comment by parkinsona** [comment]
I think I've worked out the answer to my 1st question.
The event handler will find the best match in the handlers that I have. 
In my auditTrailHandlerClass I had a handler for both UserAccountEvent & PasswordChangedEvent. When I was changing the password, it was only firing the PasswordChangedEvent. To fire the UserAccountEvent I had to remove the handler for PasswordChangedEvent. 
Is this only if the event handlers are within the same class?
PasswordChangedEvent is handled in the default EmailAccountEventsHandler and that has continued to work.
